src/charm.py

```diff
@@ -1,233 +1,294 @@
+        if not valid_pebble_plan:
+            self._handle_status_update_config(event)
             return
 
-        self.unit.status = MaintenanceStatus("restarting application")
-        container.restart(self.name)
-
-        event.set_results({"result": "application successfully restarted"})
+        check = self._container.get_check("up")
+        if check.status != CheckStatus.UP:
+            self.unit.status = MaintenanceStatus("Status check: DOWN")
+            return
 
-    def _validate(self):
-        """Validate that configuration and relations are valid and ready.
+    @log_event_handler(logger)
+    def _on_oauth_info_changed(self, event: OAuthInfoChangedEvent) -> None:
+        """Handle `oauth-info-changed` event.
 
-        Raises:
-            ValueError: in case of invalid configuration.
+        Args:
+            event: The `OAuthInfoChangedEvent` event triggered when integration data changed.
         """
-        if not self.config["upstream"]:
-            raise ValueError("missing `upstream` config")
-
-        if not self.config["cookie-secret"]:
-            raise ValueError("missing `cookie-secret` config")
-
-        if self.config["provider"] == "google" and (not self.config["client-id"] or not self.config["client-secret"]):
-            raise ValueError("`client-id` and `client-secret` config must be set for `google` provider")
-
-        if not self.config["authenticated-emails-list"] and "--email-domain=" not in self.config["additional-config"]:
-            raise ValueError(
-                "`--email-domain` must be set using `additional-config` if not setting `authenticated-emails-list`"
-            )
+        self._handle_status_update_config(event)
 
     @log_event_handler(logger)
-    def _update(self, event):
-        """Update the application configuration and replan its execution.
+    def _handle_status_update_config(self, event: HookEvent) -> None:
+        """Update the application status and configuration and restart the container.
 
         Args:
-            event: The event triggered when the relation changed.
+            event: The event triggered when the application needs to be updated.
         """
-        try:
-            self._validate()
-        except ValueError as err:
-            self.unit.status = BlockedStatus(str(err))
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to OAuth2-Proxy container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to OAuth2-Proxy container")
             return
 
-        container = self.unit.get_container(self.name)
-        if not container.can_connect():
-            self.unit.status = WaitingStatus("waiting to connect to container")
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            logger.info("Waiting for peer relation. Deferring the event.")
+            event.defer()
```

Comment:
Same as above, perhaps we should call the holistic handler when the peer relation is ready?